thunder/distributed/tensor_parallel.py

```diff
@@ -0,0 +1,333 @@
+            raise NotImplementedError("Currently thunder does not support the combination of fsdp and tensor parallel")
+
+    def __call__(
+        self,
+        prologue_trace: TraceCtx,
+        computation_trace: TraceCtx,
+        epilogue_trace: TraceCtx,
+        **kwargs,
+    ) -> tuple[TraceCtx, TraceCtx, TraceCtx]:
+        from thunder.core import prims
+        from thunder.core.pytree import tree_flatten
+        from thunder.core.transforms import visitor_transform
+
+        modules_and_thunder_modules = [
+            (bsym.args[0], bsym.output)
+            for bsym in prologue_trace.bound_symbols
+            if bsym.sym is prims.unpack_thunder_module
+        ]
+
+        if len(modules_and_thunder_modules) != 1:
+            raise NotImplementedError("cannot deal with modules other than the compiled module")
+
+        consumers = utils.consumers(computation_trace)
+        flat_args, _ = tree_flatten((computation_trace.args, computation_trace.kwargs))
+        bsyms_before_allgather: dict[BoundSymbol, PrePostProcessInterface] = {}
+        for proxy in filter(lambda p: isinstance(p, TensorProxy), flat_args):
+            for p_name in self.chunked_param_name2layer_type:
+                if p_name == proxy.name:
+                    consumer_bsym = consumers[proxy][0]
+                    if consumer_bsym not in bsyms_before_allgather:
+                        match self.chunked_param_name2layer_type[p_name]:
+                            case nn.Linear:
+                                bsyms_before_allgather[consumer_bsym] = LinearPrePostProcess(
+                                    process_group=self.process_group
+                                )
+                            case nn.Embedding:
+                                bsyms_before_allgather[consumer_bsym] = EmbeddingPrePostProcess(
+                                    num_local_embeddings=proxy.shape[0], process_group=self.process_group
+                                )
+                            case _:
+                                utils.check(
+                                    False,
+                                    lambda: f"{self.chunked_param_name2layer_type[p_name]=} is not supported",
+                                )
+        utils.check(bsyms_before_allgather, lambda: f"{bsyms_before_allgather} must not be empty")
+
+        visit = TransformVisitor(self.process_group, bsyms_before_allgather)
+        new_computation_trace = visitor_transform(
+            computation_trace,
+            visit=visit,
+            provenance="gather ouptut of column-wise tensor parallel layer",
+        )
+        if distributed_c10d.get_rank() == 0:
+            print(visit.swap_map)
+
+        return prologue_trace, new_computation_trace, epilogue_trace
+
+
+# TODO(crcrpar): Support `nn.Embedding`
+# TODO(crcrpar): Add an option to turn off output all-gather.
```

Comment:
feels like this kind of optimizations should come after this early transform as another early transform